File: AlgoProg/Semestre1/tp9/pile.c
/****************************************************************
 *  Auteur  : Xavier CRÉGUT
 *  Objectif : Implantation d'une pile de char
 ***************************************************************/

#include <assert.h>
#include <stdlib.h>
#include "pile.h"

void initialiser(Pile *p)
{
    assert(p != NULL);

    p->nb = 0;

    assert(est_vide(*p));
}

char sommet(Pile p)
{
    assert(!est_vide(p));
	return *(p.elements+p.nb-1);
}

void depiler(Pile *p)
{
    assert(p != NULL);
    assert(!est_vide(*p));
	--p->nb;
}

void empiler(Pile *p, char v)
{
    assert(p != NULL);
    assert(!est_pleine(*p));

	*(p->elements+p->nb) = v;
	++p->nb;
    assert(sommet(*p) == v);
}

bool est_vide(Pile p)
{
	return p.nb == 0;
}

bool est_pleine(Pile p)
{
	return p.nb == PILE_CAPACITE;
}

File: AlgoProg/Semestre2/tp8/Pile.java
/** Définit la structure de données Pile.
  * @author	Xavier Crégut
  */
public interface Pile<E> {

	/** Savoir si la pile est vide.
	 * @return vrai si la pile est vide
	 */
	//@ pure
	boolean estVide();

	/** Obtenir l'élément en sommet de la pile.
	  * @return l'élément en sommet de pile
	  */
	//@ requires ! this.estVide();
	//@ pure
	E sommet();

	/** Ajouter un élément au sommet de la pile.
	 * @param x élément à empiler
	 */
	//@ ensures this.sommet() == x;
	void empiler(E x);

	/** Supprimer l'élément en sommet de la pile.  */
	//@ requires ! this.estVide() ;
	void depiler();

}
